test/bench/synthetic_benchmarks.cpp

```diff
@@ -0,0 +1,269 @@
+{
+    const auto trait = instr::traits[opcode];
+    if (opcode == OP_MLOAD)
+        return InstructionCategory::mload;
+    else if (opcode == OP_MSTORE)
+        return InstructionCategory::mstore;
+    else if (opcode >= OP_PUSH1 && opcode <= OP_PUSH32)
+        return InstructionCategory::push;
+    else if (opcode >= OP_SWAP1 && opcode <= OP_SWAP16)
+        return InstructionCategory::swap;
+    else if (opcode >= OP_DUP1 && opcode <= OP_DUP16)
+        return InstructionCategory::dup;
+    else if (trait.stack_height_required == 0 && trait.stack_height_change == 0)
+        return InstructionCategory::nop;
+    else if (trait.stack_height_required == 0 && trait.stack_height_change == 1)
+        return InstructionCategory::nullop;
+    else if (trait.stack_height_required == 1 && trait.stack_height_change == 0)
+        return InstructionCategory::unop;
+    else if (trait.stack_height_required == 2 && trait.stack_height_change == -1)
+        return InstructionCategory::binop;
+    else
+        return InstructionCategory::other;
+}
+
+struct CodeParams
+{
+    evmc_opcode opcode;
+    Mode mode;
+};
+
+/// The less-than comparison operator. Needed for std::map.
+[[maybe_unused]] inline constexpr bool operator<(const CodeParams& a, const CodeParams& b) noexcept
+{
+    return std::tuple(a.opcode, a.mode) < std::tuple(b.opcode, b.mode);
+}
+
+std::string to_string(const CodeParams& params)
+{
+    return std::string{instr::traits[params.opcode].name} + '/' +
+           static_cast<char>(get_instruction_category(params.opcode)) +
+           std::to_string(static_cast<int>(params.mode));
+}
+
+/// Generates the EVM benchmark loop inner code for the given opcode and "mode".
+bytecode generate_loop_inner_code(CodeParams params)
+{
+    const auto [opcode, mode] = params;
+    const auto category = get_instruction_category(opcode);
+    switch (mode)
+    {
+    case Mode::min_stack:
+        switch (category)
+        {
+        case InstructionCategory::mload:
+            return push(0) + stack_limit * 2 * bytecode{opcode} + OP_POP;
+        case InstructionCategory::mstore:
+            return push(0) + stack_limit * (bytecode{OP_DUP1} + OP_DUP1 + bytecode{opcode}) +
+                   OP_POP;
+        case InstructionCategory::push:
+            // PUSH1 POP PUSH1 POP ...
+            return stack_limit * (push(opcode, {}) + OP_POP);
+        case InstructionCategory::swap:
+        {
+            const auto n = opcode - OP_SWAP1 + 1;
+            // DUP1 ...  SWAPn SWAPn ...  POP ...
+            // \  n   /                   \  n  /
+            return n * OP_DUP1 +                         // Required stack height for SWAPn.
+                   stack_limit * 2 * bytecode{opcode} +  // Multiple SWAPns.
+                   n * OP_POP;                           // Pop initially duplicated values.
+        }
+        case InstructionCategory::dup:
+        {
+            const auto k = opcode - OP_DUP1;  // For DUPn, k = n-1.
```

Comment:
As I understand this, DUPs, unlike SWAPs, can use the loop counter that is already on stack, and we want to minimize the overhead of all preparatory stack filling, that's why we use `n - 1`.
I think it's worth to explain in comment.

Reply:
Not exactly. SWAPn has `n` stack requirement while DUPn only `n-1`.